meta-iotqa/lib/oeqa/selftest/computervision-no-gplv3.py

```diff
@@ -0,0 +1,423 @@
+                degradedLicenses.add(license)
+            # Else we can directly use this license.
+
+    if candidateLicenses == licenses:
+        # The license set didn't change and is stable. We can go with it.
+        return licenses - degradedLicenses - set(prohibited)
+
+    # Else check the dependencies again with the new candidateLicenses. This
+    # is guaranteed to finish if the license degradation graph is acyclical.
+    return calculateConstraints(constraints, candidateLicenses, degradedLicenses)
+
+packageCache = {}
+recipeCache = {}
+rdepsCache = {}
+licenseCache = {}
+
+def getRecipe(package):
+    if package in packageCache:
+        return packageCache[package]
+
+    rRecipeProp = subprocess.check_output(["oe-pkgdata-util", "lookup-recipe", package]).decode("utf-8").strip()
+
+    packageCache[package] = rRecipeProp
+    return rRecipeProp
+
+def getPackage(recipe):
+    if recipe in recipeCache:
+        return recipeCache[recipe]
+
+    rPackageProp = subprocess.check_output(["oe-pkgdata-util", "lookup-pkg", recipe]).decode("utf-8").strip()
+
+    recipeCache[recipe] = rPackageProp
+    return rPackageProp
+
+def getRdeps(package):
+    if package in rdepsCache:
+        return rdepsCache[package]
+
+    rundepsProp = subprocess.check_output(["oe-pkgdata-util", "read-value", "RDEPENDS", package]).decode("utf-8")
+    rundeps = [token for token in rundepsProp.strip().split() if not token[0] == "(" and not token[-1] == ")"]
+    rRundeps = [getPackage(package) for package in rundeps]
+
+    rdepsCache[package] = rRundeps
+    return rRundeps
+
+def getLicenses(package):
+    if package in licenseCache:
+        return licenseCache[package]
+
+    licenseProp = subprocess.check_output(["oe-pkgdata-util", "read-value", "LICENSE", package]).decode("utf-8")
+
+    licenseCache[package] = licenseProp
+    return licenseProp
+
+def findChildren(name, whiteList, chain=[]):
+    results = []
+
+    rundeps = getRdeps(name)
+
+    for d in rundeps:
+        recipe = getRecipe(d)
+        if recipe in whiteList:
+            # Do not process whitelisted dependencies.
+            continue
+        if d in chain:
+            # Take away possible loops.
+            continue
+
+        # print(str(chain) + " -> " + d + ": " + str(rundeps))
+
+        licenses = parseLicenseString(getLicenses(d))
+
+        children = findChildren(d, whiteList, chain + [d])
+        childNode = LicenseNode(licenses, children, d)
+
+        results.append(childNode)
+
+    return results
+
+def createTree(package, whiteList=[]):
+
+    rundeps = getRdeps(package)
+
+    licenses = parseLicenseString(getLicenses(package))
+
+    children = findChildren(package, whiteList, [package])
+    root = LicenseNode(licenses, children, package)
+
+    return root
+
+class GplV3Test(oeSelfTest):
+    """GPLv3 licensing test class."""
+
+    test_image = 'refkit-image-computervision-nogplv3'
```

Comment:
Long-term I envision per-image "production" vs. "development" modes, instead of the rather limited refkit-production.inc vs. refkit-development.inc. Then refkit-image-computervision could just be re-generated in the production mode by flipping a single setting in the local build configuration of the test.

Regarding adding more tests: I think the generic license checking code should be in a meta-refkit/lib/ Python module. Then it becomes easier to reuse. Some automated tests for it would also be useful, to ensure that it really catches the errors that it is meant to catch.